libbeat/publisher/queue/diskqueue/queue.go

```diff
@@ -0,0 +1,279 @@
+	return NewQueue(logger, settings)
+}
+
+// NewQueue returns a disk-based queue configured with the given logger
+// and settings, creating it if it doesn't exist.
+func NewQueue(logger *logp.Logger, settings Settings) (queue.Queue, error) {
+	logger = logger.Named("diskqueue")
+	logger.Debugf(
+		"Initializing disk queue at path %v", settings.directoryPath())
+
+	if settings.MaxBufferSize > 0 &&
+		settings.MaxBufferSize < settings.MaxSegmentSize*2 {
+		return nil, fmt.Errorf(
+			"Disk queue buffer size (%v) must be at least "+
+				"twice the segment size (%v)",
+			settings.MaxBufferSize, settings.MaxSegmentSize)
+	}
+
+	// Create the given directory path if it doesn't exist.
+	err := os.MkdirAll(settings.directoryPath(), os.ModePerm)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't create disk queue directory: %w", err)
+	}
+
+	// Load the previous queue position, if any.
+	nextReadPosition, err := nextReadPositionFromPath(settings.stateFilePath())
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		// Errors reading / writing the position are non-fatal -- we just log a
+		// warning and fall back on the oldest existing segment, if any.
+		logger.Warnf("Couldn't load most recent queue position: %v", err)
+	}
+
+	// Index any existing data segments to be placed in segments.reading.
+	initialSegments, err := scanExistingSegments(settings.directoryPath())
+	if err != nil {
+		return nil, err
+	}
+	var nextSegmentID segmentID
+	if len(initialSegments) > 0 {
+		// Initialize nextSegmentID to the first ID after the existing segments.
+		lastID := initialSegments[len(initialSegments)-1].id
+		nextSegmentID = lastID + 1
+	}
+
+	// If any of the initial segments are older than the read position from
+	// the state file, move them directly to the acked list where they can be
+	// deleted.
+	ackedSegments := []*queueSegment{}
+	readSegmentID := nextReadPosition.segmentID
+	for len(initialSegments) > 0 && initialSegments[0].id < readSegmentID {
+		ackedSegments = append(ackedSegments, initialSegments[0])
+		initialSegments = initialSegments[1:]
+	}
+
+	// If the next read position is older than all existing segments, advance
+	// it to the beginning of the first one.
+	if len(initialSegments) > 0 && readSegmentID < initialSegments[0].id {
+		nextReadPosition = queuePosition{segmentID: initialSegments[0].id}
+	}
+
+	// We wait for four goroutines: core loop, reader loop, writer loop,
+	// deleter loop.
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(4)
+
+	// The helper loops all have an input channel with buffer size 1, to ensure
+	// that the core loop can never block when sending a request (the core
+	// loop never sends a request until receiving the response from the
+	// previous one, so there is never more than one outstanding request for
+	// any helper loop).
+
+	readerLoop := &readerLoop{
+		settings: &settings,
+
+		requestChan:  make(chan readerLoopRequest, 1),
+		responseChan: make(chan readerLoopResponse),
```

Comment:
I'm not sure which aspect you're getting at here -- are you saying the reader loop itself should create its response channel? Or that the request channel should be owned by the core loop? (Fwiw, the response channels are never closed, and the request channels are closed by the core loop to signal the loop to shut down.)

Reply:
I was referring to the pipeline pattern (although I know it's not a long one here).

For reference: https://blog.golang.org/pipelines

Feel free to ignore this idea if there is no big gain in this case.

Reply:
My tentative feeling is that this case doesn't fit the pipeline pattern well enough -- these aren't stages in a stream of transformed values, they're 1-1 request-response APIs where we always know exactly which channels have a request in progress and where all state is owned by the central call site. Since the output channels don't need to be closed they also don't particularly need to be "owned" by anybody, so the question is mostly about where it's most readable to initialize them.

In that spirit though, I made helper functions to create each of the loop structures from the queue settings, which I think makes the main initialization function a lot less crowded, so see if you are happier with it this way :-)